routing/probability_estimator.go

```diff
@@ -0,0 +1,161 @@
+	// the weighted average calculation below. When the apriori weight
+	// approaches 1, the apriori factor goes to infinity. It will heavily
+	// outweigh any observations that have been collected.
+	aprioriFactor := 1/(1-p.aprioriWeight) - 1
+
+	// Calculate a weighted average consisting of the apriori probability
+	// and historical observations. This is the part that incentives nodes
+	// to make sure that all (not just some) of their channels are in good
+	// shape. Senders will steer around nodes that have shown a few
+	// failures, even though there may be many channels still untried.
+	//
+	// If there is just a single observation and the apriori weight is 0,
+	// this single observation will totally determine the node probability.
+	// The node probability is returned for all other channels of the node.
+	// This means that one failure will lead to the success probability
+	// estimates for all other channels being 0 too. The probability for the
+	// channel that was tried will not even recover, because it is
+	// recovering to the node probability (which is zero). So one failure
+	// effectively prunes all channels of the node forever. This is the most
+	// aggressive way in which we can penalize nodes and unlikely to yield
+	// good results in a real network.
+	probabilitiesTotal := p.aprioriHopProbability * aprioriFactor
+	totalWeight := aprioriFactor
+
+	for _, result := range results {
+		age := now.Sub(result.timestamp)
+
+		switch {
+		// Weigh success with a constant high weight of 1. There is no
+		// decay.
+		case result.success:
+			totalWeight++
+			probabilitiesTotal += p.prevSuccessProbability
+
+		// Weigh failures in accordance with their age. The base
+		// probability of a failure is considered zero, so nothing needs
+		// to be added to probabilitiesTotal.
+		case amt >= result.minPenalizeAmt:
+			totalWeight += p.getWeight(age)
+		}
+	}
+
+	return probabilitiesTotal / totalWeight
+}
+
+// getWeight calculates a weight in the range [0, 1] that should be assigned to
+// a payment result. Weight follows an exponential curve that starts at 1 when
+// the result is fresh and asymptotically approaches zero over time. The rate at
+// which this happens is controlled by the penaltyHalfLife parameter.
+func (p *probabilityEstimator) getWeight(age time.Duration) float64 {
+	exp := -age.Hours() / p.penaltyHalfLife.Hours()
+	return math.Pow(2, exp)
+}
+
+// getPairProbability estimates the probability of successfully traversing to
+// toNode based on historical payment outcomes for the from node. Those outcomes
+// are passed in via the results parameter.
+func (p *probabilityEstimator) getPairProbability(
+	now time.Time, results map[route.Vertex]timedPairResult,
+	toNode route.Vertex, amt lnwire.MilliSatoshi) float64 {
+
+	// Getter for the node probability that is used for untried channels.
+	getNodeProbability := func() float64 {
+		return p.getNodeProbability(now, results, amt)
```

Comment:
I don't think this function closure really adds much, other than introducing 3 extra lines in this diff.

Reply:
It was an attempt to deduplicate code. But was in doubt already myself about it. Unrolled.